packages/gatsby/src/internal-plugins/query-runner/query-runner.js

```diff
@@ -58,15 +60,57 @@ module.exports = async (pageOrLayout, component) => {
   }
   const resultJSON = JSON.stringify(result)
   const resultHash = md5(resultJSON)
-  const resultPath = joinPath(
-    program.directory,
-    `.cache`,
-    `json`,
-    pageOrLayout.jsonName
-  )
-
-  if (resultHashes[resultPath] !== resultHash) {
-    resultHashes[resultPath] = resultHash
-    await fs.writeFile(resultPath, resultJSON)
+
+  if (resultHashes[pageOrLayout.jsonName] !== resultHash) {
+    resultHashes[pageOrLayout.jsonName] = resultHash
+    const programType = program._[0]
+
+    // In production, write file to public/static/d/ folder.
+    if (programType === `build`) {
+      const dataPath = `${generatePathChunkName(
+        pageOrLayout.jsonName
+      )}-${resultHash}`
+
+      const resultPath = path.join(
+        program.directory,
+        `public`,
+        `static`,
+        `d`,
+        `${dataPath}.json`
+      )
+      await fs.writeFile(resultPath, resultJSON)
+
+      if (!pageOrLayout.path) {
+        store.dispatch({
+          type: `SET_LAYOUT_DATA_PATH`,
+          payload: {
+            id: pageOrLayout.id,
+            dataPath,
+          },
+        })
+      } else {
+        store.dispatch({
+          type: `SET_PAGE_DATA_PATH`,
+          payload: {
+            path: pageOrLayout.path,
+            dataPath,
+          },
+        })
+      }
+
+      return
+    }
+
+    // In development queue up results until a client is available
+    // push subsequent results to client
+    if (programType === `develop`) {
+      const result = {}
```

Comment:
Should this be `const` if it's mutated on the following line? I would have expected it to throw an error, but I guess it doesn't if the code works.

Reply:
`const` only means that we can't reassign `result` reference - we can't later do `result = 'something else'`, but we can mutate object itself